File: Math/src/Vector3.cpp
#include <math.h>

#include "Math/Vector3.h"

namespace Math
{
    float Vector3::Sum(const Vector3& a) { return a.x + a.y + a.z; }
    float Vector3::Dot(const Vector3& a, const Vector3& b) { return a.x * b.x + a.y * b.y + a.z * b.z; }
    float Vector3::LengthSquared(const Vector3& a) { return a.x * a.x + a.y * a.y + a.z * a.z; }
    float Vector3::Length(const Vector3& a) { return sqrt(a.x * a.x + a.y * a.y + a.z * a.z); }

    Vector3 Vector3::Normalize(const Vector3&a)
    {
        if (a == Vector3::Zero)
        {
            return a;
        }
        else
        {
            return a / Vector3::Length(a);
        }
    }

    Vector3 Vector3::Cross(const Vector3& a, const Vector3& b)
    {
        return Vector3(
            a.y * b.z - a.z * b.y,
            a.z * b.x - a.x * b.z,
            a.x * b.y - a.y * b.x
        );
    }

    const Vector3 Vector3::Zero = Vector3();
    const Vector3 Vector3::One = Vector3(1.0f, 1.0f, 1.0f);
    const Vector3 Vector3::Right = Vector3(1.0f);
    const Vector3 Vector3::Up = Vector3(0.0f, 1.0f);
    const Vector3 Vector3::Forward = Vector3(0.0f, 0.0f, 1.0f);

    // Vector Operators
    bool operator==(const Vector3& a, const Vector3& b) { return a.x == b.x && a.y == b.y && a.z == b.z; }
    bool operator!=(const Vector3& a, const Vector3& b) { return a.x != b.x || a.y != b.y || a.z != b.z; }

    Vector3 operator+(const Vector3& a, const Vector3& b) { return Vector3(a.x + b.x, a.y + b.y, a.z + b.z); }
    Vector3 operator-(const Vector3& a, const Vector3& b) { return Vector3(a.x - b.x, a.y - b.y, a.z - b.z); }
    Vector3 operator-(const Vector3& a) { return Vector3(-a.x, -a.y, -a.z); }

    Vector3 operator*(const float& a, const Vector3& b) { return Vector3(a * b.x, a * b.y, a * b.z); }
    Vector3 operator*(const Vector3& a, const float& b) { return Vector3(a.x * b, a.y * b, a.z * b); }
    Vector3 operator*(const Vector3& a, const Vector3& b) { return Vector3(a.x * b.x, a.y * b.y, a.z * b.z); }

    Vector3 operator/(const Vector3& a, const float& b) { return Vector3(a.x / b, a.y / b, a.z / b); }

    Vector3& operator+=(Vector3& a, const Vector3& b)
    { 
        a.x += b.x;
        a.y += b.y;
        a.z += b.z;
        return a;
    }

    std::ostream& operator<<(std::ostream& out, const Vector3& v) { return out << "Vector3(" << v.x << ", " << v.y << ", " << v.z << ")"; }
}

File: Assets/Scripts/Core/Math/Vector3.cs
using System;

namespace NoEngine
{
    public struct Vector3 : IEquatable<Vector3>
    {
        public static readonly Vector3 Zero = new Vector3(0, 0, 0);
        public static readonly Vector3 Right = new Vector3(1, 0, 0);
        public static readonly Vector3 Up = new Vector3(0, 1, 0);
        public static readonly Vector3 Forward = new Vector3(0, 0, 1);
        public static readonly Vector3 One = new Vector3(1, 1, 1);

        public float x, y, z;

        public Vector3(float x, float y, float z)
        {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        public static Vector3 Normalize(Vector3 vector)
        {
            if (vector == Vector3.Zero) return vector;

            float magnitude = (float) Math.Sqrt((double) (vector.x * vector.x + vector.y * vector.y + vector.z * vector.z));
            return new Vector3(vector.x / magnitude, vector.y / magnitude, vector.z / magnitude);
        }

        public bool Equals(Vector3 other) { return this == other; }
        public override bool Equals(object other)
        {
            if (!(other is Vector3)) return false;
            return this.Equals((Vector3) other);
        }

        public override int GetHashCode()
        {
            return x.GetHashCode() ^ y.GetHashCode() ^ z.GetHashCode();
        }

        #region Operators
        public static bool operator ==(Vector3 a, Vector3 b) { return a.x == b.x && a.y == b.y && a.z == b.z; }
        public static bool operator !=(Vector3 a, Vector3 b) { return a.x != b.x || a.y != b.y || a.z != b.z; }
        public static Vector3 operator +(Vector3 a, Vector3 b) { return new Vector3(a.x + b.x, a.y + b.y, a.z + b.z); }
        public static Vector3 operator -(Vector3 a, Vector3 b) { return new Vector3(a.x - b.x, a.y - b.y, a.z - b.z); }
        public static Vector3 operator *(float a, Vector3 b) { return new Vector3(a * b.x, a * b.y, a * b.z); }
        #endregion
    }
}